File: app/src/main/java/com/github/vatbub/scoreboard/GameTableRecyclerViewAdapter.java
package com.github.vatbub.scoreboard;

import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class GameTableRecyclerViewAdapter extends RecyclerView.Adapter<GameTableViewHolder> {
    private GameManager.Game game;
    private RecyclerView parent;
    private Set<GameTableViewHolder> mBoundViewHolders = new HashSet<>();
    private int lastLineColumnWidth;
    private MainActivity mainActivity;

    public GameTableRecyclerViewAdapter(RecyclerView parent, GameManager.Game game, MainActivity mainActivity) {
        this.game = game;
        this.parent = parent;
        setMainActivity(mainActivity);
        registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeChanged(int positionStart, int itemCount) {
                updateLineNumbers();
            }

            @Override
            public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
                updateLineNumbers();
            }

            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                updateLineNumbers();
            }

            @Override
            public void onItemRangeRemoved(int positionStart, int itemCount) {
                updateLineNumbers();
            }

            @Override
            public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
                updateLineNumbers();
            }

            @Override
            public void onChanged() {
                updateLineNumbers();
            }
        });
    }

    public Set<GameTableViewHolder> getAllBoundViewHolders() {
        return Collections.unmodifiableSet(mBoundViewHolders);
    }

    @Override
    public GameTableViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View rowView = LayoutInflater.from(parent.getContext()).inflate(R.layout.scoreboard_row, parent, false);
        return new GameTableViewHolder(rowView);
    }

    @Override
    public void onBindViewHolder(final GameTableViewHolder holder, int position) {
        mBoundViewHolders.add(holder);

        holder.setLineNumber(holder.getAdapterPosition() + 1);
        if (lastLineColumnWidth != 0)
            holder.getLineNumberTextView().setWidth(lastLineColumnWidth);

        holder.getDeleteRowButton().setOnClickListener(v -> {
            if (holder.getAdapterPosition() == RecyclerView.NO_POSITION) return;
            getGame().removeScoreLineAt(holder.getAdapterPosition());
            getMainActivity().updateLineNumberWidth();
            notifyItemRemoved(holder.getAdapterPosition());
        });

        final List<Long> scoreLine = getGame().getScoreLineAt(holder.getAdapterPosition());
        holder.getScoreHolderLayout().removeAllViews();

        for (int i = 0; i < scoreLine.size(); i++) {
            final long score = scoreLine.get(i);
            final EditText editText = new EditText(holder.getView().getContext());
            editText.setInputType(EditorInfo.TYPE_CLASS_NUMBER);
            editText.setText(String.format(editText.getContext().getString(R.string.main_table_score_template), score));
            final int finalI = i;
            editText.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {

                }

                @Override
                public void afterTextChanged(Editable s) {
                    try {
                        long newScore = 0;
                        if (s.length() > 0)
                            newScore = Long.valueOf(s.toString());

                        scoreLine.set(finalI, newScore);
                        getGame().modifyScoreLineAt(holder.getAdapterPosition(), scoreLine);
                    } catch (NumberFormatException e) {
                        Toast.makeText(holder.getView().getContext(), R.string.max_input_length_reached_toast, Toast.LENGTH_LONG).show();
                        s.delete(s.length() - 1, s.length());
                    }
                }
            });

            LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, 1);
            editText.setLayoutParams(layoutParams);

            holder.getScoreHolderLayout().addView(editText);
        }
    }

    @Override
    public void onViewRecycled(GameTableViewHolder holder) {
        mBoundViewHolders.remove(holder);
    }

    public void updateLineNumbers() {
        for (int childCount = getParent().getChildCount(), i = 0; i < childCount; ++i) {
            GameTableViewHolder holder = (GameTableViewHolder) getParent().getChildViewHolder(getParent().getChildAt(i));
            if (holder.getAdapterPosition() != RecyclerView.NO_POSITION)
                holder.setLineNumber(holder.getAdapterPosition() + 1);
        }
    }

    public void updateColumnWidths(int columnWidth) {
        lastLineColumnWidth = columnWidth;
        for (GameTableViewHolder holder : getAllBoundViewHolders()) {
            holder.getLineNumberTextView().setWidth(columnWidth);
            // holder.getLineNumberTextView().requestLayout();
        }
    }

    @Override
    public int getItemCount() {
        return getGame().getScoreCount();
    }

    public GameManager.Game getGame() {
        return game;
    }


    public RecyclerView getParent() {
        return parent;
    }

    public MainActivity getMainActivity() {
        return mainActivity;
    }

    public void setMainActivity(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
    }
}

File: app/src/main/java/com/github/vatbub/scoreboard/GameTableRecyclerViewAdapter.kt
package com.github.vatbub.scoreboard

import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Toast
import com.github.vatbub.scoreboard.data.Game
import java.util.*

class GameTableRecyclerViewAdapter(val parent: RecyclerView, val game: Game, val mainActivity: MainActivity) : RecyclerView.Adapter<GameTableViewHolder>() {
    private val mBoundViewHolders = HashSet<GameTableViewHolder>()
    private var lastLineColumnWidth: Int = 0

    val allBoundViewHolders: Set<GameTableViewHolder>
        get() = Collections.unmodifiableSet(mBoundViewHolders)

    init {
        registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onItemRangeChanged(positionStart: Int, itemCount: Int) = updateLineNumbers()
            override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) = updateLineNumbers()
            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) = updateLineNumbers()
            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) = updateLineNumbers()
            override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) = updateLineNumbers()
            override fun onChanged() = updateLineNumbers()
        })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GameTableViewHolder {
        val rowView = LayoutInflater.from(parent.context).inflate(R.layout.scoreboard_row, parent, false)
        return GameTableViewHolder(rowView, true)
    }

    override fun onBindViewHolder(holder: GameTableViewHolder, position: Int) {
        mBoundViewHolders.add(holder)

        holder.setLineNumber(holder.adapterPosition + 1)
        if (lastLineColumnWidth != 0)
            holder.lineNumberTextView.width = lastLineColumnWidth

        holder.deleteRowButton.setOnClickListener {
            if (holder.adapterPosition == RecyclerView.NO_POSITION)
                return@setOnClickListener
            game.removeScoreLineAt(holder.adapterPosition)
            mainActivity.updateLineNumberWidth()
            mainActivity.renderSumRow()
            mainActivity.renderLeaderboard()
            notifyItemRemoved(holder.adapterPosition)
        }

        val scoreLine = game.getScoreLineAt(holder.adapterPosition)
        val scoreLineCopy = scoreLine.toMutableList()
        holder.scoreHolderLayout.removeAllViews()

        scoreLine.forEachIndexed { index, score ->
            val editText = EditText(holder.view.context)
            editText.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)
            editText.inputType = EditorInfo.TYPE_CLASS_NUMBER
            editText.hint = editText.context.getString(R.string.main_table_score_template, 0)
            editText.setHorizontallyScrolling(false)
            editText.maxLines = AppConfig.maxLinesForEnterText
            val scoreString = editText.context.getString(R.string.main_table_score_template, score)
            if (score != 0L) editText.setText(scoreString)

            editText.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(editable: Editable) {
                    try {
                        var newScore: Long = 0
                        if (editable.isNotEmpty())
                            newScore = editable.toString().toLong()

                        scoreLineCopy[index] = newScore
                        game.modifyScoreLineAt(holder.adapterPosition, scoreLineCopy)
                        mainActivity.renderSumRow()
                        mainActivity.renderLeaderboard()
                    } catch (e: NumberFormatException) {
                        Toast.makeText(holder.view.context, R.string.max_input_length_reached_toast, Toast.LENGTH_LONG).show()
                        editable.delete(editable.length - 1, editable.length)
                    }

                }
            })

            holder.scoreHolderLayout.addView(editText)
        }
    }

    override fun onViewRecycled(holder: GameTableViewHolder?) {
        mBoundViewHolders.remove(holder)
    }

    fun updateLineNumbers() {
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val holder = parent.getChildViewHolder(parent.getChildAt(i)) as GameTableViewHolder
            if (holder.adapterPosition != RecyclerView.NO_POSITION)
                holder.setLineNumber(holder.adapterPosition + 1)
        }
    }

    fun updateColumnWidths(columnWidth: Int) {
        lastLineColumnWidth = columnWidth
        allBoundViewHolders.forEach { it.lineNumberTextView.width = columnWidth }
    }

    override fun getItemCount(): Int {
        return game.scoreCount
    }
}
